gcc/rust/util/optional.h

```diff
@@ -0,0 +1,2460 @@
+
+///
+// optional - An implementation of std::optional with extensions
+// Written in 2017 by Sy Brand (tartanllama@gmail.com, @TartanLlama)
+//
+// Documentation available at https://tl.tartanllama.xyz/
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to the
+// public domain worldwide. This software is distributed without any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software. If not, see
+// <http://creativecommons.org/publicdomain/zero/1.0/>.
+///
+
+#ifndef TL_OPTIONAL_HPP
+#define TL_OPTIONAL_HPP
+
+#define TL_OPTIONAL_VERSION_MAJOR 1
+#define TL_OPTIONAL_VERSION_MINOR 1
+#define TL_OPTIONAL_VERSION_PATCH 0
+
+// TODO: Are these okay?
+#include <exception>
```

Comment:
A bootstrap build will show if it works, I say we should include this into the rust-session-manager and merge it and see how MArks build farm copes with it before merging the other code so we can fix anything we find

Reply:
IIRC, gcc is built without exceptions https://gcc.gnu.org/codingconventions.html#Exceptions

Reply:
I think it's okay to modify the file so it includes "rust-system.h" instead of the poisoned headers. As mentioned on the #2317 PR, I'll split up the commit into adding the original file and then modifying it.